docs/spec/concepts.rst

```diff
@@ -3,97 +3,174 @@
 Type system concepts
 ====================
 
-.. _`union-types`:
+Static, dynamic, and gradual typing
+-----------------------------------
+
+A **statically-typed** programming language runs a type-checker before running
+a program. The program is required to be well-typed according to the language's
+type system. The type system assigns a type to all expressions in the language
+and verifies that their uses obey the typing rules. Normally, a program that is
+not well-typed (i.e., one that contains a type error) will not run. Java and
+C++ are examples of statically-typed object-oriented languages.
+
+A **dynamically-typed** programming language does not run a type-checker before
+running a program. Instead, it will check the types of values before performing
+operations on them at runtime. This is not to say that the language is
+"untyped". Values at runtime have a type and their uses obey typing rules. Not
+every operation will be checked, but certain primitive operations in the
+language such as attribute access or arithmetic will be.  Python was
+historically a dynamically-typed language.
+
+**Gradual typing** is the name for a specific way to combine static and dynamic
+typing.  It allows mixing static and dynamic checking at a fine level of
+granularity.  For instance, in Python, individual variables, parameters, and
+return values can be either statically or dynamically typed.  Data structures
+such as objects can have a mix of static and dynamic checking.  As an example,
+a Python dictionary could choose to have static checking of the key type but
```

Comment:
It's odd to me to describe this as a choice by the dictionary itself, rather than by the user. I'm not sure I agree with the earlier sentence "in Python, individual variables, parameters, and return values can be either statically or dynamically typed". In Python the language all values are dynamically typed; we're describing a static type system, but that system is separate from the language itself.

Reply:
This is a place where I'd prefer to be a little more careful about using the word "value" - I don't like saying that values have types.

I know in this case "return value" is being used as a synonym of "returned expression" and "dictionary" is being used as a synonym for "dictionary term", but I've seen discussion from users who haven't really internalized that static types exist on terms of code, not on the values flowing through the runtime themselves.

As a result I'd like us to to be pretty careful about our language.

Reply:
@stroxler in a sense values do have types.  If static types denote sets of values, then a value has the type of the static type denotations it is a member of.

We more or less need that idea if we want to talk about "soundness".  The type system is sound if, whenever we have an expression `e` of type `T`, then the evaluation of `e` will produce a value that agrees with `T` in some sense.

Reply:
I've attempted a re-wording of this paragraph, taking into account the concerns mentioned here. Here's my effort:

> **Gradual typing** is the name for a specific way to combine static and dynamic
typing. It allows mixing static and dynamic checking at a fine level of
granularity. For instance, in type-annotated Python, individual variables,
parameters, and returns can optionally be given a static type. Data structures
such as objects can have a mix of static and dynamic-only checking. As an
example, a Python dictionary could be annotated to have static checking of the
key type but only dynamic checking of the value type.